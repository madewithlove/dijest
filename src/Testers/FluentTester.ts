import { ComponentFactory } from '../component';
import { mount, render } from 'enzyme';
import matchesSnapshot from '../helpers/matchesSnapshot';
import toJson from 'enzyme-to-json';

export enum TestType {
    Any,
    Component,
}

export default class FluentTester {
    /**
     * Middlewares to apply to the tested party
     */
    middlewares: Function[] = [];

    /**
     * The test runner to use
     */
    tester: jest.It = it;

    /**
     * The test's description
     */
    description: string;

    /**
     * The value being tested
     */
    tested: any;

    /**
     * The type of value being tested
     */
    type: TestType = TestType.Any;

    /**
     * The expected value
     */
    expected?: any;

    setDescription(description: string) {
        this.description = description;

        return this;
    }

    setTested(value: any, type: TestType = TestType.Any) {
        this.tested = value;
        this.type = type;

        return this;
    }

    setExpected(value: any) {
        this.expected = value;

        return this;
    }

    /**
     * Get the tested value in a workable form for callbacks
     */
    getWorkableValue(): any {
        switch (this.type) {
            case TestType.Component:
                return mount(this.tested);

            default:
                return this.tested;
        }
    }

    /**
     * Get the tested value in serializable form for snapshots
     */
    getSerializedValue(): string | object {
        switch (this.type) {
            case TestType.Component:
                return toJson(render(this.tested));

            default:
                return this.tested;
        }
    }

    /**
     * Run the tests
     */
    run() {
        this.tester(this.description, () => {
            if (typeof this.expected !== 'undefined') {
                expect(this.tested).toBe(this.expected);
            } else {
                expect(this.getSerializedValue()).toMatchSnapshot();
            }
        });
    }

    /**
     * Run multiple tests with the given context
     */
    then(callback: Function) {
        this.tester(this.description, () => {
            callback(this.getWorkableValue(), matchesSnapshot);
        });
    }
}
