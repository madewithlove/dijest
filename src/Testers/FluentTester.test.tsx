import FluentTester from "./FluentTester";

new FluentTester()
    .setDescription("can build test with fluent interface")
    .setTested(true)
    .setExpected(true)
    .run();

new FluentTester()
    .setDescription("can use snapshots")
    .setTested({ foo: "bar" })
    .run();
