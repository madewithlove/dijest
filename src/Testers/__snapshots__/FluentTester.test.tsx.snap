// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can use snapshots 1`] = `
Object {
  "foo": "bar",
}
`;
